File: src/46.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        if len(nums) == 0:
            return []
        
        res = []
        visited = [0 for i in range(len(nums))]
        # numSet = set()
        # self.dfs(nums, numSet, [], res)
        self.dfs(nums, visited, [], res)
        return res
    
    def dfs(self, nums, visited, temp, res):
        if len(temp) == len(nums):
            res.append([] + temp)
        else:
            for i in range(len(nums)):
                # if nums[i] in numSet:
                #     continue
                if visited[i] == 1:
                    continue
                    
                temp.append(nums[i])
                # numSet.add(nums[i])
                visited[i] = 1
                self.dfs(nums, visited, temp, res)
                # numSet.remove(nums[i])
                visited[i] = 0
                temp.pop()
        

File: src/46.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        if (nums.size() == 0) {
            return res;
        }
        
        vector<int> temp;
        vector<int> visited(nums.size());
        for (int i = 0; i < nums.size(); i ++) {
            visited[i] = 0;
        }
        
        dfs(nums, res, visited, temp);
        return res;
        
    }
    
    void dfs(vector<int> &nums, vector<vector<int>> &res, vector<int> visited, vector<int> temp) {
        if (temp.size() == nums.size()) {
            res.push_back(temp);
            return;
        } else {
            for (int i = 0; i < nums.size(); i ++) {
                if (visited[i] == 1) {
                    continue;
                }
                temp.push_back(nums[i]);
                visited[i] = 1;
                dfs(nums, res, visited, temp);
                visited[i] = 0;
                temp.pop_back();
            }
        }
    }
};